### PR TITLE
[SCB-150] Add Status(success/failed) dimension to operation call count and tps

### PR DESCRIPTION
**Pull request by zhengyangyong**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

Add Status(success/failed) dimension to operation call count and tps,now they will split into three metrics like:
servicecomb.instance.producer.producerCall.tps.{Status=all}
servicecomb.instance.producer.producerCall.tps.{Status=success}
servicecomb.instance.producer.producerCall.tps.{Status=failed}
servicecomb.instance.producer.producerCall.total.{Status=all}
servicecomb.instance.producer.producerCall.total.{Status=success}
servicecomb.instance.producer.producerCall.total.{Status=failed}
*consumerCall also applied*

In prometheus will use [Label](https://prometheus.io/docs/concepts/data_model/) show this dimension:
servicecomb_calculator_calculatorRestEndpoint_calculate_producer_producerCall_tps{Status="success",instance="localhost:9696",job="servicecomb"}
servicecomb_calculator_calculatorRestEndpoint_calculate_producer_producerCall_tps{Status="failed",instance="localhost:9696",job="servicecomb"}
servicecomb_calculator_calculatorRestEndpoint_calculate_producer_producerCall_tps{Status="all",instance="localhost:9696",job="servicecomb"}

**Update**
now support setting output level of status dimension :
```yaml
servicecomb:
  metrics:
    dimension:
      #status dimension
      status:
        # success_failed  : only output success & failed (default)
        # code_group : output 1xx,2xx,3xx,4xx,5xx and xxx
        # code       : output detail code like 200,404,500... etc
        output_level: success_failed
```

### CODE REVIEW

**Comment by zhengyangyong** [comment]
**Don't Merge This PR before our package change from 'io.servicecomb' to apache completed,but review is welcome**

**Comment by wujimin** [comment]
maybe it's better add statusCode dimension, not succ/fail

**Comment by zhengyangyong** [comment]
Package Name Updated

**Comment by zhengyangyong** [comment]
maybe it's better add statusCode dimension, not succ/fail
@wujimin I just think it's necessary split them so detail ? @WillemJiang @seanyinx  what's your suggestion thanks

**Comment by WillemJiang** [comment]
There could be lots's of dimensions if we create the counter per status code.   My suggestion is we only provide 2xx, 3xx, 4xx, 5xx status code for it. BTW, we can take the Spring Boot actuator as an example, we don't need lots detail metrics, we just need to tell the administrator something is wrong,  and you can find more information by looking up the access log.

**Comment by wujimin** [comment]
2xx/3xx/4xx/5xx almost equals statusCode
and 3xx need to support special logic: 304 belongs success

i thinks, for metrics kernel data, we care for statusCode
but for data published interface, we can provide different model for different scenes

**Comment by zhengyangyong** [comment]
I think I can add a switch named *servicecomb.metrics.dimension.http_code*,default value is **false**,if user enable it,detail metrics will be record like : 
servicecomb.instance.producer.producerCall.tps.{Status=all}
servicecomb.instance.producer.producerCall.tps.{Status=success}
servicecomb.instance.producer.producerCall.tps.{Status=failed}
servicecomb.instance.producer.producerCall.tps.{Status=200}
servicecomb.instance.producer.producerCall.tps.{Status=404}
...

**Comment by zhengyangyong** [comment]
I also think most of the time user would like use metrics outputs draw curves directly and don't want do any more aggregation,only provide detail items may not good idea.

**Comment by wujimin** [comment]
not "only detail items "
detail is our kernel data, by this data, we can summarize everything we want.

File: metrics/metrics-common/src/main/java/org/apache/servicecomb/metrics/common/MetricsDimension.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+  public static final String DIMENSION_STATUS_FAILED = "failed";
+
+  public static String[] getDimensionOptions(String dimension) {
+    if (DIMENSION_STATUS.equals(dimension)) {
+      return new String[] {MetricsDimension.DIMENSION_STATUS_ALL,
+          MetricsDimension.DIMENSION_STATUS_SUCCESS,
+          MetricsDimension.DIMENSION_STATUS_FAILED};
+    }
+    throw new ServiceCombException("illegal dimension key : " + dimension);
```
Comment (seanyinx) [inline metrics/metrics-common/src/main/java/org/apache/servicecomb/metrics/common/MetricsDimension.java]: if there's only one dimension, there's no need to pass in the parameter. if there are multiple, returning an empty array is preferable

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/utils/MonitorUtils.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      Map<String, String> tagMap = new HashMap<>();
+      for (Tag tag : tags) {
+        //we don't need servo internal type tag for metrics
+        if (!"type".equals(tag.getKey())) {
+          tagMap.put(tag.getKey(), tag.getValue());
+        }
+      }
+      return tagMap;
+    }
+    return null;
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/utils/MonitorUtils.java]: returning Collections.emptyMap() is preferable

File: metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/MetricsCollector.java [hunk 3]
```diff
@@ -56,24 +58,76 @@ public MetricsCollector(DataSource dataSource) {
...
+    List<Sample> samples = new ArrayList<>();
+    samples.addAll(convertMetricValues(registryMetric.getInstanceMetric().getSystemMetric().toMap()));
+    samples.addAll(convertConsumerMetric(registryMetric.getInstanceMetric().getConsumerMetric()));
+    samples.addAll(convertCallMetric(registryMetric.getInstanceMetric().getConsumerMetric().getConsumerCall()));
+    samples.addAll(convertProducerMetric(registryMetric.getInstanceMetric().getProducerMetric()));
+    samples.addAll(convertCallMetric(registryMetric.getInstanceMetric().getProducerMetric().getProducerCall()));
+    familySamples.add(new MetricFamilySamples("Instance Level", Type.UNTYPED, "Instance Level Metrics", samples));
+
+    if (registryMetric.getConsumerMetrics().size() != 0) {
+      samples = new ArrayList<>();
```
Comment (seanyinx) [inline metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/MetricsCollector.java]: use a local list in this scope is better

File: metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/MetricsCollector.java [hunk 4]
```diff
@@ -56,24 +58,76 @@ public MetricsCollector(DataSource dataSource) {
...
+        samples.addAll(convertCallMetric(metric.getConsumerCall()));
+      }
+      familySamples.add(new MetricFamilySamples("Consumer Side", Type.UNTYPED, "Consumer Side Metrics", samples));
     }
-    for (Entry<String, ProducerInvocationMetric> producerMetric : registryMetric.getProducerMetrics().entrySet()) {
-      familySamples
-          .add(getFamilySamples(producerMetric.getKey() + " Producer Side", producerMetric.getValue().toMap()));
+
+    if (registryMetric.getProducerMetrics().size() != 0) {
+      samples = new ArrayList<>();
```
Comment (seanyinx) [inline metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/MetricsCollector.java]: same as above

File: metrics/metrics-common/src/main/java/org/apache/servicecomb/metrics/common/DoubleMetricValue.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      @JsonProperty("value") Double value,
+      @JsonProperty("dimensions") Map<String, String> dimensions) {
+    super(key, value, dimensions);
+  }
+
+  private DoubleMetricValue merge(DoubleMetricValue value) {
+    if (this.getKey().equals(value.getKey())) {
+      return new DoubleMetricValue(this.getKey(), this.getValue() + value.getValue(), this.getDimensions());
+    }
+    throw new ServiceCombException("unable merge different key values,source key :" + value.getKey() +
```
Comment (seanyinx) [inline metrics/metrics-common/src/main/java/org/apache/servicecomb/metrics/common/DoubleMetricValue.java]: any place to catch this exception? we don't want to interrupt user business logic

Comment (zhengyangyong) [reply metrics/metrics-common/src/main/java/org/apache/servicecomb/metrics/common/DoubleMetricValue.java]: Done

File: metrics/metrics-common/src/main/java/org/apache/servicecomb/metrics/common/MetricsDimension.java (hunk 1 above)
Comment (zhengyangyong) [reply metrics/metrics-common/src/main/java/org/apache/servicecomb/metrics/common/MetricsDimension.java]: We will support more dimensions later, and dimension key must correct

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/utils/MonitorUtils.java (hunk 2 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/utils/MonitorUtils.java]: Done

File: metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/MetricsCollector.java (hunk 3 above)
Comment (zhengyangyong) [reply metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/MetricsCollector.java]: Done

File: metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/MetricsCollector.java (hunk 4 above)
Comment (zhengyangyong) [reply metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/MetricsCollector.java]: Done

**Comment by zhengyangyong** [comment]
now support setting output level of status dimension :
```yaml
servicecomb:
  metrics:
    dimension:
      #status dimension
      status:
        # success_failed  : only output success & failed (default)
        # code_group : output 1xx,2xx,3xx,4xx,5xx and xxx
        # code       : output detail code like 200,404,500... etc
        output_level: success_failed
```

**Comment by wujimin** [comment]
will this make our logic too complex?
our metrics will not record history data, only some summaries, there is no memory problem?
if kernel data is about statusCode, then metrics is very simple
just when publish data, need to publish different model like: succ/fail,    2xx/3xx/4xx......

**Pull request merged**